{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.3"
    },
    "colab": {
      "name": "2019227.ipynb",
      "provenance": []
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ll8uuKFaqNST"
      },
      "source": [
        "#2A\n",
        "#SEQ  1    SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDTVYCPRHVICTAEDML   50          ~~~~\n",
        "#STR       TTTT     HHHHHH EEEEEETTEEEEEEEETTEEEEEGGGG  HHHHH               ~~~~\n",
        "#REM                                                                        ~~~~\n",
        "#REM                .         .         .         .         .               ~~~~\n",
        "#SEQ  51   NPNYEDLLIRKSNHSFLVQAGNVQLRVIGHSMQNCLLRLKVDTSNPKTPK  100          ~~~~\n",
        "#STR          HHHHHHH  GGG EEEETTEEE EEEEEEETTEEEEEE   TTTT                 ~~~~\n",
        "#REM                                                                        ~~~~\n",
        "#REM                .         .         .         .         .               ~~~~\n",
        "#SEQ  101  YKFVRIQPGQTFSVLACYNGSPSGVYQCAMRPNHTIKGSFLNGSCGSVGF  150          ~~~~\n",
        "#STR             TTTEEEEEEEEETTEEEEEEEEEETTTT B    TTTTTTTEE                ~~~~"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "r1zasDr3qNSZ",
        "outputId": "a274c498-5045-4099-dcc4-0314887fb023"
      },
      "source": [
        "import operator as op\n",
        "\n",
        "\n",
        "#initiating Chou Fasman Data for alpha and beta listed below as a dictionary \n",
        "#initiating a dictionary for amino acids\n",
        "aminoAcid = {}\n",
        "aminoAcid['A']      = \"alanine\" \n",
        "aminoAcid['R']      = \"arginine\"\n",
        "aminoAcid['N']      = \"aspartic acid\"\n",
        "aminoAcid['D']      = \"aspargine\"\n",
        "aminoAcid['C']      = \"cysteine\"\n",
        "aminoAcid['E']      = \"glutamic acid\"\n",
        "aminoAcid['Q']      = \"glutamine\"\n",
        "aminoAcid['G']      = \"glycine\"\n",
        "aminoAcid['H']      = \"histidine\"\n",
        "aminoAcid['I']      = \"isoleucine\"\n",
        "aminoAcid['L']      = \"leucine\"\n",
        "aminoAcid['K']      = \"lysine\"\n",
        "aminoAcid['M']      = \"methonine\"\n",
        "aminoAcid['F']      = \"phenylalanine\"\n",
        "aminoAcid['P']      = \"proline\"\n",
        "aminoAcid['S']      = \"serine\"\n",
        "aminoAcid['T']      = \"theoreonine\"\n",
        "aminoAcid['W']      = \"tryptophan\"\n",
        "aminoAcid['Y']      = \"tyrosine\"\n",
        "aminoAcid['V']      = \"valine\"\n",
        "\n",
        "alphaValues = {}\n",
        "#initiating a dictionary for weights of alpha \n",
        "alphaValues[\"glutamic acid\"]=1.53 \n",
        "alphaValues[\"alanine\"]=1.45  \n",
        "alphaValues[\"leucine\"]=1.34 \n",
        "alphaValues[\"histidine\"]=1.24\n",
        "alphaValues[\"methonine\"]=1.20 \n",
        "alphaValues[\"glutamine\"]=1.17\n",
        "alphaValues[\"tryptophan\"]=1.14\n",
        "alphaValues[\"valine\"]=1.14 \n",
        "alphaValues[\"phenylalanine\"]=1.12\n",
        "alphaValues[\"lysine\"]=1.07\n",
        "alphaValues[\"isoleucine\"]=1.00 \n",
        "alphaValues[\"aspartic acid\"]=0.98\n",
        "alphaValues[\"theoreonine\"]=0.82\n",
        "alphaValues[\"serine\"]=0.79\n",
        "alphaValues[\"arginine\"]=0.79\n",
        "alphaValues[\"cysteine\"]=0.77\n",
        "alphaValues[\"aspargine\"]=0.73\n",
        "alphaValues[\"tyrosine\"]=0.61\n",
        "alphaValues[\"proline\"]=0.59\n",
        "alphaValues[\"glycine\"]=0.53\n",
        "\n",
        "betaValues = {}\n",
        "#initiating a dictionary for weights of beta strand \n",
        "betaValues[\"methonine\"]=1.67 \n",
        "betaValues[\"valine\"]=1.65 \n",
        "betaValues[\"isoleucine\"]=1.60 \n",
        "betaValues[\"cysteine\"]=1.30 \n",
        "betaValues[\"tyrosine\"]=1.29 \n",
        "betaValues[\"phenylalanine\"]=1.28 \n",
        "betaValues[\"glutamine\"]=1.23 \n",
        "betaValues[\"leucine\"]=1.22  \n",
        "betaValues[\"theoreonine\"]=1.20 \n",
        "betaValues[\"tryptophan\"]=1.19 \n",
        "betaValues[\"alanine\"]=0.97 \n",
        "betaValues[\"arginine\"]=0.90\n",
        "betaValues[\"glycine\"]=0.81 \n",
        "betaValues[\"aspartic acid\"]=0.80\n",
        "betaValues[\"lysine\"]=0.74\n",
        "betaValues[\"serine\"]=0.72\n",
        "betaValues[\"histidine\"]=0.71\n",
        "betaValues[\"aspargine\"]=0.65\n",
        "betaValues[\"proline\"]=0.62\n",
        "betaValues[\"glutamic acid\"]=0.26 \n",
        "\n",
        "inputProteinSequence = \"SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDTVYCPRHVICTAEDMLNPNYEDLLIRKSNHSFLVQAGNVQLRVIGHSMQNCLLRLKVDTSNPKTPKYKFVRIQPGQTFSVLACYNGSPSGVYQCAMRPNHTIKGSFLNGSCGSVGF\"\n",
        "\n",
        "#initiated an empty Helix list as well as an empty beta strand list \n",
        "Helix = op.mul([0],len(inputProteinSequence))\n",
        "Strand = op.mul([0],len(inputProteinSequence))\n",
        "\n",
        "#this function is responsible for fetching all the alpha helices in the given sequence\n",
        "def alphaHelix():\n",
        "\tstart = 0 \n",
        "\twhile op.lt(op.add(start , 6) , len(inputProteinSequence)): #parameters for alpha being the window = 6 \n",
        "\t\tcheckIndex = 0 #this basically keeps track of the count for all the alpha helices in counts of 6 \n",
        "\t\tfor i in range(start,op.add(start , 6)):\n",
        "\t\t\tif op.ge(alphaValues[aminoAcid[inputProteinSequence[i]]],1):\n",
        "\t\t\t\tcheckIndex=op.add(checkIndex, 1)\n",
        "\t\tif op.ge(checkIndex,4):\n",
        "\t\t\tcheckForAlpha(start,op.add(start , 6))\n",
        "\t\tstart = op.add(start , 1)\n",
        "\n",
        "#this code is responsible checking with respect to right adjustment index as well as the left adjustment index for the possible alpha helix sequence \n",
        "def checkForAlpha(start,end):\n",
        "\trightIndexing=op.truediv(sum([alphaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start, 3):op.add(start, 7)]]),4)\n",
        "\tline = start + 7\n",
        "\twhile op.and_(op.lt(line , len(inputProteinSequence)) , op.ge(rightIndexing , 1)):\n",
        "\t\tline= op.add(line, 1)\n",
        "\t\trightIndexing = op.truediv(sum([alphaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start, 3):op.add(start, 7)]]),4)\n",
        "\n",
        "\tleftIndexing= op.truediv(sum([alphaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start, 1):op.add(start,3)]]),4)\n",
        "\n",
        "\tbooline = start - 1 \n",
        "\twhile op.and_(op.gt(booline , 0) ,op.ge(leftIndexing ,1)):\n",
        "\t\tbooline= op.sub(booline , 1)\n",
        "\t\tleftIndexing = op.truediv(sum([alphaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start,1):op.add(start,3)]]),4)\n",
        "\n",
        "\tfor i in range(start,end):\n",
        "\t\tHelix[i]=True\n",
        "\n",
        "#this function is responsible for fetching all the beta strands in the given protein sequence\n",
        "def betaStrand():\n",
        "\tstart = 0\n",
        "\twhile op.lt(op.add(start,5) , len(inputProteinSequence)):\n",
        "\t\tcheckIndex=0  #this basically keeps track of the count for all the beta strange in counts of 5 \n",
        "\t\tfor i in range(start,op.add(start,5)):\n",
        "\t\t\tif op.ge(betaValues[aminoAcid[inputProteinSequence[i]]],1):\n",
        "\t\t\t\tcheckIndex=op.add(checkIndex, 1)\n",
        "\t\tif op.ge(checkIndex,3):\n",
        "\t\t\tcheckForBeta(start,op.add(start,5))\n",
        "\t\tstart=op.add(start,1)\n",
        "\n",
        "#this code is responsible checking with respect to right adjustment index as well as the left adjustment index for the possible beta strande sequences\n",
        "def checkForBeta(start,end):\n",
        "\trightIndexing=op.truediv(sum([betaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(end,3):op.add(end,1)]]),4)\n",
        "\n",
        "\twhile op.and_(op.lt(end , len(inputProteinSequence)) , op.ge(rightIndexing ,1) ): \n",
        "\t\tend = op.add(end, 1)\n",
        "\t\trightIndexing = op.truediv(sum([betaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(end,3):op.add(end,1)]]),4)\n",
        "\tleftIndexing= op.truediv(sum([betaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start,1):op.add(start,3)]]),4)\n",
        "\n",
        "\twhile op.and_(op.gt(start , 0) , op.ge(leftIndexing ,1)):\n",
        "\t\tstart = op.sub(start,1)\n",
        "\t\tleftIndexing = op.truediv(sum([betaValues[aminoAcid[x]] for x in inputProteinSequence[op.sub(start,1):op.add(start,3)]]),4)\n",
        "\tfor i in range(start,end):\n",
        "\t\tStrand[i]=True\n",
        "\n",
        "#this function is responsible for generating the secondary sequence structre \n",
        "def secondarySequenceStructure():\n",
        "\ti=0\n",
        "\tsecondary_sequence = \"\"\n",
        "\twhile op.lt(i , len(inputProteinSequence)):\n",
        "\t\tif op.and_(op.eq(Helix[i],False) ,op.eq(Strand[i],False)):\n",
        "\t\t\tj = i \n",
        "\t\t\twhile j < len(inputProteinSequence) and op.eq(Helix[j] , False ) and op.eq(Strand[j] , False): \n",
        "\t\t\t\tsecondary_sequence += 'T' #the Turns is signified with T, reporting the empty regions \n",
        "\t\t\t\tj=op.add(j,1 )\n",
        "\t\t\ti = j\n",
        "\n",
        "\t\telif op.and_( op.eq(Helix[i] , True) , op.eq(Strand[i] , False)):\n",
        "\t\t\tj = i \n",
        "\t\t\twhile op.le(j , len(inputProteinSequence))and op.eq(Helix[j] , True) and op.eq(Strand[j] , False): \n",
        "\t\t\t\tsecondary_sequence += 'H' #this is responsible for filling the possible alpha helices \n",
        "\t\t\t\tj=op.add(j,1 )\n",
        "\t\t\ti=j\n",
        "\n",
        "\t\telif op.and_(op.eq(Helix[i],False) , op.eq(Strand[i],True)):\n",
        "\t\t\tj = i \n",
        "\t\t\twhile op.lt(j ,len(inputProteinSequence))  and op.eq(Helix[j] ,False) and op.eq(Strand[j] , True): \n",
        "\t\t\t\tsecondary_sequence += 'S' # this loop is responsible for filling the possible beta strands \n",
        "\t\t\t\tj=op.add(j,1 )\n",
        "\t\t\ti=j\n",
        "\n",
        "\t\telif op.and_(op.eq(Helix[i],True) , op.eq(Strand[i],True)): \n",
        "\t\t\tj = i \n",
        "\t\t\tHelix_sum = 0 \n",
        "\t\t\tStrand_sum = 0 \n",
        "\t\t\twhile op.lt(j ,len(inputProteinSequence)) and op.eq(Helix[j] , True) and op.eq(Strand[j] , True):\n",
        "\t\t\t\tHelix_sum  = op.add(Helix_sum,alphaValues[aminoAcid[inputProteinSequence[j]]])\n",
        "\t\t\t\tStrand_sum = op.add(Strand_sum,betaValues[aminoAcid[inputProteinSequence[j]]])\n",
        "\t\t\t\tj=op.add(j,1 )\n",
        "\t\t\tfor z in range(i,j):\n",
        "\t\t\t\tif op.ge(Helix_sum , Strand_sum):\n",
        "\t\t\t\t\tsecondary_sequence = op.add(secondary_sequence,'H') \n",
        "\t\t\t\telse:\n",
        "\t\t\t\t\tsecondary_sequence = op.add(secondary_sequence,'S')\n",
        "\t\t\ti = j \n",
        "\treturn secondary_sequence\n",
        "alphaHelix() #calling the alpha function responsible for generating the alpha helices \n",
        "betaStrand() #calling the alpha function responsible for generating the beta strands \n",
        "print(secondarySequenceStructure())\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "THHHHHHHHHTHSSSSSSSSSSSSSSHHHHHHSSSSSSSSHHHHHHHHHHHTTHHHHHHHHHHSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSTTTTHSSSSSSSSSSSSSSSSSSSSSTTSSSSSSSSSSTTTHHHHHHTTTTTTTTT\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Mkyiy8dtqNSc"
      },
      "source": [
        "#SEQ  1    SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDTVYCPRHVICTAEDML   50          ~~~~\n",
        "#STR       TTTT     HHHHHH EEEEEETTEEEEEEEETTEEEEEGGGG  HHHHH               ~~~~\n",
        "#REM                                                                        ~~~~\n",
        "#REM                .         .         .         .         .               ~~~~\n",
        "#SEQ  51   NPNYEDLLIRKSNHSFLVQAGNVQLRVIGHSMQNCLLRLKVDTSNPKTPK  100          ~~~~\n",
        "#STR          HHHHHHH  GGG EEEETTEEE EEEEEEETTEEEEEE   TTTT                 ~~~~\n",
        "#REM                                                                        ~~~~\n",
        "#REM                .         .         .         .         .               ~~~~\n",
        "#SEQ  101  YKFVRIQPGQTFSVLACYNGSPSGVYQCAMRPNHTIKGSFLNGSCGSVGF  150          ~~~~\n",
        "#STR             TTTEEEEEEEEETTEEEEEEEEEETTTT B    TTTTTTTEE                ~~~~\n",
        "\n",
        "#My output\n",
        "#THHHHHHHHHTHSSSSSSSSSSSSSSHHHHHHSSSSSSSSHHHHHHHHHHHTTHHHHHHHHHHSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSTTTTHSSSSSSSSSSSSSSSSSSSSSTTSSSSSSSSSSTTTHHHHHHTTTTTTTTT\n",
        "\n",
        "#We notice that both of the outputs are different with respect to the way the algo was even created, empty\n",
        "#regions are present in ChouFasman algo run by Stride. \n",
        "#We can also observe that there are gaps in their output which in turn we have account for with respect \n",
        "#to the T character. In my algorithm, I have assigned turn ‘T’ wherever alpha and beta was not coming across,\n",
        "#whereas in the Stride Output, this condition was not kept in mind, for generating this ‘T’ in their output, \n",
        "#they have implemented another logic, owing to which we observe a difference in both their outputs and my output. \n",
        "#Chou and Fasman was used a logic where they took care of indexes where no alpha or beta was present ,which I \n",
        "#have otherwise done using turn ‘T’.\n",
        "\n",
        "\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "45fncFd6qNSc"
      },
      "source": [
        "#2B\n",
        "\n",
        "#1)In our code, we have only assigned 3 symbols H , S and T each denoting Alpha Helices , Beta Strands and Turns \n",
        "#respectively whereas if we notice in the ChouFasman Algo deduced by the 2 scientists there are more variables \n",
        "#allotted. There are the G, E and B variables as well in this algo, which we haven’t taken into account. \n",
        "#2)The E character denotes strand in the Stride output. I used S for the same. \n",
        "#3)The B character denotes bridge in the Stride output. \n",
        "#4)The G character denotes 310helix with respect to the stride output, which we haven't taken into account. \n",
        "#5)The C character accounts for the Coils. \n",
        "#6)The chou fasman method was developed in the 1970s, and it is an advanced method devoted for higher bioinformatics\n",
        "#concepts, unlike the algorithm developed by me which is a layman’s interpretation of the same algo created \n",
        "#by Chou and Fasman. \n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xH_L7y2TqNSd"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Csciz1duqNSd"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}